pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/BaseInstanceSelector.java

```diff
@@ -18,69 +18,194 @@
 
-  private final String _tableNameWithType;
   private final BrokerMetrics _brokerMetrics;
+  private final String _tableNameWithType;
+  private final ZkHelixPropertyStore<ZNRecord> _propertyStore;
   protected final AdaptiveServerSelector _adaptiveServerSelector;
 
   // These 4 variables are the cached states to help accelerate the change processing
   private Set<String> _enabledInstances;
   private Map<String, List<String>> _segmentToOnlineInstancesMap;
-  private Map<String, List<String>> _segmentToOfflineInstancesMap;
   private Map<String, List<String>> _instanceToSegmentsMap;
+  private Map<String, SegmentState> _newSegmentStates;
 
-  // These 2 variables are needed for instance selection (multi-threaded), so make them volatile
-  private volatile Map<String, List<String>> _segmentToEnabledInstancesMap;
-  private volatile Set<String> _unavailableSegments;
+  // _segmentStateSnapshot is needed for instance selection (multi-threaded), so make them volatile
+  private volatile SegmentStateSnapshot _segmentStateSnapshot;
+  private Clock _clock;
 
   BaseInstanceSelector(String tableNameWithType, BrokerMetrics brokerMetrics,
-      @Nullable AdaptiveServerSelector adaptiveServerSelector) {
+      @Nullable AdaptiveServerSelector adaptiveServerSelector, ZkHelixPropertyStore<ZNRecord> propertyStore) {
+    this(tableNameWithType, brokerMetrics, adaptiveServerSelector, propertyStore, Clock.systemUTC());
+  }
+
+  // Test only for clock injection.
+  BaseInstanceSelector(String tableNameWithType, BrokerMetrics brokerMetrics,
+      @Nullable AdaptiveServerSelector adaptiveServerSelector, ZkHelixPropertyStore<ZNRecord> propertyStore,
+      Clock clock) {
     _tableNameWithType = tableNameWithType;
     _brokerMetrics = brokerMetrics;
     _adaptiveServerSelector = adaptiveServerSelector;
+    _newSegmentStates = new HashMap<>();
+    _propertyStore = propertyStore;
+    _clock = clock;
+  }
+
+  // Get the segment where the ideal state hasn't converged with external view and which doesn't have error instance.
+  private static List<String> getPotentialNewSegments(IdealState idealState, ExternalView externalView,
+      Set<String> onlineSegments) {
+    List<String> potentialNewSegments = new ArrayList<>();
+    Map<String, Map<String, String>> externalViewAssignment = externalView.getRecord().getMapFields();
+    for (Map.Entry<String, Map<String, String>> entry : idealState.getRecord().getMapFields().entrySet()) {
+      String segment = entry.getKey();
+      // Only track online segments
+      if (!onlineSegments.contains(segment)) {
+        continue;
+      }
+      Map<String, String> idealStateInstanceStateMap = entry.getValue();
+      // Segments with missing external view are considered as new.
+      Map<String, String> externalViewInstanceStateMap =
+          externalViewAssignment.getOrDefault(segment, Collections.emptyMap());
+      List<String> onlineInstance = new ArrayList<>();
+      boolean couldBeNewSegment = true;
```

Comment:
Do we need this flag? Since we break on L163, we probably can remove the check on `couldBeNewSegment` and still get the same logic.